### PR TITLE
Update files as part of the Scaffold.Execute

### LINKED ISSUE #1226: [Scaffold enhancement] File interface
# Context

## Current situation

The current `File` interface is as follows:

https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/input/input.go#L141-L145

The single method is not descriptive enough and does different tasks:
1. Sets default values
2. Returns a non-descriptive `Input` struct which contains:
    - Two fields needed to prepare the template: `Path` and `TemplateBody`
    - Several other fileds that are not used

Note: the `Input` struct also works as a parent class for mostly all `File`-implementing structs so that they can be injected some data from the scaffold. By mostly I mean all except the `PROJECT` file, which is an special-cased yaml file which doesn't have an embedded `Input`.

## `GetInput` usage

<details><summary>I have tracked <code>GetInput</code> usages.</summary>

### `pkg/scaffold/project.go`
Lines 101 and 106 for v1, 192 and 197 for v2:
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L101
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L106
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L192
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L197
The created variables are used in lines 113 and 127 for v1, and 204 and 218 for v2:
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L113
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L127
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L204
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project.go#L218

Only the path is needed. Setting the default values (which is also done by `GetInput`) is not needed here except for the path itself.

### `pkg/scaffold/scaffold.go`
Line 229:
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/scaffold.go#L229
Which is later used in line 235 and 238:
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/scaffold.go#L235
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/scaffold.go#L238
The usage in `doTemplate` are in lines 278, 291 and 292:
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/scaffold.go#L278
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/scaffold.go#L291-L292

The path is needed in lines 235, 291 and 292. The template body is needed in line 278. The side-effect of setting the default values is important here as the files are used to execute the templates in line 284:
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/scaffold.go#L284

### `pkg/scaffold/project/project_test.go`
Line 49:
https://github.com/kubernetes-sigs/kubebuilder/blob/03f70fb1a0f7521f383a08e41562124021161baf/pkg/scaffold/project/project_test.go#L48-L52
It is used to set the default values.</details>

# Proposal

A new interface is proposed:

```go
type File interface {
	GetPath() string
	GetTemplateBody() string
	SetDefaults() error
}
```

<details><summary>Method usage</summary>

- `GetPath` would be used in lines 113, 127, 204 and 218 of `pkg/scaffold/project.go` and in lines 235, 291 and 292 of `pkg/scaffold/scaffold.go`.
- `GetTemplateBody` would be used in line 278 of `pkg/scaffold/scaffold.go`.
- `SetDefaults` would be called before line 284 of `pkg/scaffold/scaffold.go` and in line 49 of `pkg/scaffold/project/project_test.go`.
- All the `GetInput` calls with their error handlings could be removed.</details>

This feature is part of a bigger change tracked in #1218

/kind feature

### LINKED ISSUE #1358: Template mixins
# Current situation

`template.Input` struct is used to inject fields from the scaffolder to the templates such as the domain, the repository or the boilerplate. It is basically a mixin that, when embedded to a template, provides inject functionality.

Having a single mixin for all the fields is neither memory efficient, as most templates only need one or two of these fields, nor performant, as we are injecting fields we are not using.

# Proposal

Instead of having a single mixin for all the injectable fields, several specialized mixins would be used instead.

Each mixin would be implemented by an interface and a struct with a single field and a single method to set this field. For the following example, a made up field `MadeUp` of type `string` is going to be used.

```go
// HasMadeUp allows a made-up to be used on a template
type HasmadeUp interface {
	// InjectMadeUp sets the template made-up
	InjectMadeUp(string)
}

// MadeUpMixin provides templates with a injectable made-up field
type MadeUpMixin struct {
	MadeUp string
}

// InjectMadeUp implements HasMadeUp
func (m *MadeUpMixin) InjectMadeUp(madeUp string) {
	if m.MadeUp == "" {
		m.MadeUp = madeUp
	}
}
```

The interface would be used to inject this field:

```go
func (u Universe) InjectInto(t file.Template) {
	/* ... */
	// Inject made-up
	if u.MadeUp != "" {
		if templateWithMadeUp, hasMadeUp := t.(file.HasMadeUp); hasMadeUp {
			templateWithMadeUp.InjectMadeUp(u.MadeUp)
		}
	}
}
```

And the mixin would be used for those templates that needs that field:

```go
var _ file.Template = &MadeUpTemplate{}

// MadeUpTemplate scaffolds a made up file
type MadeUp struct {
	file.Input
	file.MadeUpMixin
}
```

/kind feature
This belongs to the scaffold enhancement tracked by the umbrella issue #1218

### PR DESCRIPTION
**Pull request by Adirio**
# Description
Update already scaffolded files inside `Scaffold.Execute` following the same flow used for new files (injecting fields, validating, creating a model, piping to plugins and persisting it to disk).

Coverage reports 72.1% in `pkg/scaffold/internal/filesystem` (the remaining 27.9% are calls to the filesystem that cannot be further tested) and 93.8% in `pkg/scaffold/internal/machinery`, aiming to fulfill #1342.

# Motivation
File updates (i.e., inserting code fragments to already scaffolded files) was not needed for v1. When v2 came, the feature was needed and it was added outside of the main `Scaffold.Execute`. Providing `Scaffold.Execute` the capability of inserting those code fragments achieves a better work flow.

# Implementation details
This PR is made of 3 commits. The first two were submitted as separate PR previously:
1. #1359
2. #1227
I suggest reviewing them commit by commit (GitHub offers the possibility of showing only changes in one commit) as it affects more than 3500 LoC.

Regarding to the last commit, the main changes are found in `pkg/scaffold/internal/machinery/scaffold.go` and the interfaces in `pkg/model/file/interfaces.go`.

This PR is scoped under #1218
Closes #1226
Closes #1227
Closes #1358
Closes #1359

### CODE REVIEW

File: pkg/model/file/template.go [hunk 1]
```diff
@@ -20,16 +20,40 @@ import (
...
-type Input struct {
+// Template is a scaffoldable file template
+type Template interface {
+	// GetPath returns the path to the file location
+	GetPath() string
+	// GetBody returns the template body
+	GetBody() string
+	// GetIfExistsAction returns the behavior when creating a file that already exists
+	GetIfExistsAction() IfExistsAction
+	// SetTemplateDefaults returns the TemplateMixin for creating a scaffold file
```
Comment (DirectXMan12) [inline pkg/model/file/template.go]: this godoc doesn't make sense -- it returns error

File: pkg/model/file/marker.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	panic(fmt.Errorf("unknown file extension: '%s', expected '.go' or '.yaml'", ext))
+}
+
+// String implements Stringer
+func (m Marker) String() string {
+	return m.comment + prefix + m.value
+}
+
+type CodeFragments []string
```
Comment (DirectXMan12) [inline pkg/model/file/marker.go]: needs godoc

File: pkg/scaffold/internal/filesystem/errors.go [hunk 3]
```diff
@@ -24,6 +24,38 @@ import (
...
+	path string
+	err  error
+}
+
+func (e fileExistsError) Error() string {
+	return fmt.Sprintf("failed to check if %s exists: %v", e.path, e.err)
+}
+
+// IsFileExistsError checks if the returned error is because the file could not be checked for existence
+func IsFileExistsError(e error) bool {
```
Comment (DirectXMan12) [inline pkg/scaffold/internal/filesystem/errors.go]: we should eventually switch these all to use `errors.As`

File: pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go [hunk 4]
```diff
@@ -56,6 +58,51 @@ func (f *SuiteTest) Validate() error {
...
+`
+)
+
+// GetCodeFragments implements file.Inserter
+func (f *SuiteTest) GetCodeFragments() file.CodeFragmentsMap {
+	fragments := make(file.CodeFragmentsMap, 2)
+
+	// Generate import code fragments
+	imports := make([]string, 0)
+	imports = append(imports, fmt.Sprintf(apiImportCodeFragment, f.Resource.ImportAlias, f.Resource.Package))
```
Comment (DirectXMan12) [inline pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go]: `imports := []string{fmt.Sprintf(...)}`?

File: pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go [hunk 5]
```diff
@@ -56,6 +58,51 @@ func (f *SuiteTest) Validate() error {
...
+
+	// Generate add scheme code fragments
+	addScheme := make([]string, 0)
+	addScheme = append(addScheme, fmt.Sprintf(addschemeCodeFragment, f.Resource.ImportAlias))
+
+	// Only store code fragments in the map if the slices are non-empty
+	if len(imports) != 0 {
```
Comment (DirectXMan12) [inline pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go]: don't we know that these are non-empty due the the append?

File: pkg/scaffold/internal/filesystem/errors.go (hunk 3 above)
Comment (Adirio) [reply pkg/scaffold/internal/filesystem/errors.go]: I agree, I submited a PR for that and didnt get any answer, it can be done after this gets merged.

**Comment by mengqiy** [comment]
> Closes #1226
Closes #1227
Closes #1358
Closes #1359

2 of them are PRs. Why?

File: pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go (hunk 4 above)
Comment (Adirio) [reply pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go]: I did it this way on purpose, to allow easy addition and deletion of code fragments. Do you have a strong opinion on this?

File: pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go (hunk 5 above)
Comment (Adirio) [reply pkg/scaffold/internal/templates/v2/controller/controller_suitetest.go]: Same as https://github.com/kubernetes-sigs/kubebuilder/pull/1375#discussion_r382309187

**Comment by Adirio** [comment]
> 2 of them are PRs. Why?

The two first commits in this PR are those PRs. Merging this would be like merging the three.

File: pkg/model/file/marker.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"fmt"
+	"path/filepath"
+)
+
+const prefix = "+kubebuilder:scaffold:"
+
+var commentsByExt = map[string]string{
+	// TODO(v3): machine-readable comments should not have spaces by Go convention. However,
```
Comment (mengqiy) [inline pkg/model/file/marker.go]: Can you create an issue to track everything related to `TODO(v3)`?

Comment (Adirio) [reply pkg/model/file/marker.go]: #1382, #1383, and #1384 raised. Could you add them to the corresponding milestone?

**Pull request merged**